File: FirebaseRemoteConfig/Sources/FIRRemoteConfigComponent.m
/*
 * Copyright 2019 Google
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#import "FirebaseRemoteConfig/Sources/FIRRemoteConfigComponent.h"

#import "FirebaseCore/Extension/FirebaseCoreInternal.h"
#import "FirebaseRemoteConfig/Sources/Private/FIRRemoteConfig_Private.h"
#import "FirebaseRemoteConfig/Sources/RCNConfigContent.h"
#import "FirebaseRemoteConfig/Sources/RCNConfigDBManager.h"
#import "Interop/Analytics/Public/FIRAnalyticsInterop.h"

@implementation FIRRemoteConfigComponent

// Because Component now need to register two protocols (provider and interop), we need a way to
// return the same component instance for both registered protocol, this singleton pattern allow us
// to return the same component object for both registration callback.
static NSMutableDictionary<NSString *, FIRRemoteConfigComponent *> *_componentInstances = nil;

+ (FIRRemoteConfigComponent *)getComponentForApp:(FIRApp *)app {
  @synchronized(_componentInstances) {
    // need to init the dictionary first
    if (!_componentInstances) {
      _componentInstances = [[NSMutableDictionary alloc] init];
    }
    if (![_componentInstances objectForKey:app.name]) {
      _componentInstances[app.name] = [[self alloc] initWithApp:app];
    }
    return _componentInstances[app.name];
  }
  return nil;
}

+ (void)clearAllComponentInstances {
  @synchronized(_componentInstances) {
    [_componentInstances removeAllObjects];
  }
}

/// Default method for retrieving a Remote Config instance, or creating one if it doesn't exist.
- (FIRRemoteConfig *)remoteConfigForNamespace:(NSString *)remoteConfigNamespace {
  if (!remoteConfigNamespace) {
    // TODO: Throw an error? Return nil? What do we want to do?
    return nil;
  }

  // Validate the required information is available.
  FIROptions *options = self.app.options;
  NSString *errorPropertyName;
  if (options.googleAppID.length == 0) {
    errorPropertyName = @"googleAppID";
  } else if (options.GCMSenderID.length == 0) {
    errorPropertyName = @"GCMSenderID";
  } else if (options.projectID.length == 0) {
    errorPropertyName = @"projectID";
  }

  if (errorPropertyName) {
    NSString *const kFirebaseConfigErrorDomain = @"com.firebase.config";
    [NSException
         raise:kFirebaseConfigErrorDomain
        format:@"%@",
               [NSString
                   stringWithFormat:
                       @"Firebase Remote Config is missing the required %@ property from the "
                       @"configured FirebaseApp and will not be able to function properly. Please "
                       @"fix this issue to ensure that Firebase is correctly configured.",
                       errorPropertyName]];
  }

  FIRRemoteConfig *instance = self.instances[remoteConfigNamespace];
  if (!instance) {
    FIRApp *app = self.app;
    id<FIRAnalyticsInterop> analytics =
        app.isDefaultApp ? FIR_COMPONENT(FIRAnalyticsInterop, app.container) : nil;
    instance = [[FIRRemoteConfig alloc] initWithAppName:app.name
                                             FIROptions:app.options
                                              namespace:remoteConfigNamespace
                                              DBManager:[RCNConfigDBManager sharedInstance]
                                          configContent:[RCNConfigContent sharedInstance]
                                              analytics:analytics];
    self.instances[remoteConfigNamespace] = instance;
  }

  return instance;
}

/// Default initializer.
- (instancetype)initWithApp:(FIRApp *)app {
  self = [super init];
  if (self) {
    _app = app;
    _instances = [[NSMutableDictionary alloc] initWithCapacity:1];
  }
  return self;
}

#pragma mark - Lifecycle

+ (void)load {
  // Register as an internal library to be part of the initialization process. The name comes from
  // go/firebase-sdk-platform-info.
  [FIRApp registerInternalLibrary:self withName:@"fire-rc"];
}

#pragma mark - Interoperability

+ (NSArray<FIRComponent *> *)componentsToRegister {
  FIRComponent *rcProvider = [FIRComponent
      componentWithProtocol:@protocol(FIRRemoteConfigProvider)
        instantiationTiming:FIRInstantiationTimingAlwaysEager
              creationBlock:^id _Nullable(FIRComponentContainer *container, BOOL *isCacheable) {
                // Cache the component so instances of Remote Config are cached.
                *isCacheable = YES;
                return [FIRRemoteConfigComponent getComponentForApp:container.app];
              }];

  // Unlike provider needs to setup a hard dependency on remote config, interop allows an optional
  // dependency on RC
  FIRComponent *rcInterop = [FIRComponent
      componentWithProtocol:@protocol(FIRRemoteConfigInterop)
        instantiationTiming:FIRInstantiationTimingAlwaysEager
              creationBlock:^id _Nullable(FIRComponentContainer *container, BOOL *isCacheable) {
                // Cache the component so instances of Remote Config are cached.
                *isCacheable = YES;
                return [FIRRemoteConfigComponent getComponentForApp:container.app];
              }];
  return @[ rcProvider, rcInterop ];
}

#pragma mark - Remote Config Interop Protocol

- (void)registerRolloutsStateSubscriber:(id<FIRRolloutsStateSubscriber>)subscriber
                                    for:(NSString * _Nonnull)namespace {
  FIRRemoteConfig *instance = [self remoteConfigForNamespace:namespace];
  [instance addRemoteConfigInteropSubscriber:subscriber];
}

@end

File: FirebaseRemoteConfig/Sources/FIRRemoteConfigComponent.swift
import Foundation
import FirebaseCore
class FIRRemoteConfigComponent :NSObject, FIRRemoteConfigProvider, FIRRemoteConfigInterop {

    static var componentInstances = [String : FIRRemoteConfigComponent]()
    
    var app : FIRApp?
    var instances : NSMutableDictionary<NSString,FIRRemoteConfig> = [:]

    static func getComponent(app: FIRApp) -> FIRRemoteConfigComponent? {
        @synchronized(componentInstances) {
            if (componentInstances.isEmpty) {
              componentInstances = [String : FIRRemoteConfigComponent]()
            }
            if (componentInstances[app.name] == nil) {
                componentInstances[app.name] = FIRRemoteConfigComponent(app: app)
            }
            return componentInstances[app.name]
        }
        return nil
    }

    static func clearAllComponentInstances() {
        @synchronized(componentInstances) {
          componentInstances.removeAll()
        }
    }

    func remoteConfigForNamespace(firebaseNamespace: String) -> FIRRemoteConfig? {
        if firebaseNamespace.isEmpty {
            return nil
        }

        // Validate the required information is available.
        guard let options = self.app?.options else {
            fatalError("The 'options' property was not available for this configuration")
        }

        if options.googleAppID.isEmpty {
          fatalError("Firebase Remote Config is missing the required googleAppID property from the " +
                     "configured FirebaseApp and will not be able to function properly. Please " +
                     "fix this issue to ensure that Firebase is correctly configured.")
        }
        
        if options.GCMSenderID.isEmpty {
          fatalError("Firebase Remote Config is missing the required GCMSenderID property from the " +
                     "configured FirebaseApp and will not be able to function properly. Please " +
                     "fix this issue to ensure that Firebase is correctly configured.")
        }

        if options.projectID.isEmpty {
            fatalError("Firebase Remote Config is missing the required projectID property from the " +
                       "configured FirebaseApp and will not be able to function properly. Please " +
                       "fix this issue to ensure that Firebase is correctly configured.")
        }
        
        var instance : FIRRemoteConfig? = self.instances[firebaseNamespace]
        if (instance == nil) {
            let appName : String = self.app?.name ?? ""
            let dbManager : RCNConfigDBManager = RCNConfigDBManager.sharedInstance()
            let configContent = RCNConfigContent.sharedInstance()
            let googleAppID = options.googleAppID ?? ""

            instance = FIRRemoteConfig(appName: appName,
                                         FIROptions: options,
                                         namespace: FIRNamespace, DBManager: dbManager,
                                         configContent: configContent, analytics: nil)
            
            self.instances[firebaseNamespace] = instance;
        }

        return instance
    }

    init(app: FIRApp) {
        self.app = app;
        self.instances = [:]
    }

    static func load() {
        // Register as an internal library to be part of the initialization process. The name comes
        // from go/firebase-sdk-platform-info.
        FIRApp.registerInternalLibrary(self, withName: "fire-rc")
    }
    
    
    
    static func componentsToRegister() -> [FIRComponent] {
      let rcProvider = FIRComponent(protocol: FIRRemoteConfigProvider.self,
                                      instantiationTiming: .alwaysEager) {
        (container, isCacheable) -> AnyObject in
        // Cache the component so instances of Remote Config are cached.
        var isCacheable = true
        return FIRRemoteConfigComponent.getComponent(app: container.app) as Any
      }
        
      let rcInterop = FIRComponent(protocol: FIRRemoteConfigInterop.self,
                                      instantiationTiming: .alwaysEager) {
        (container, isCacheable) -> AnyObject in
        // Cache the component so instances of Remote Config are cached.
        var isCacheable = true
        return FIRRemoteConfigComponent.getComponent(app: container.app) as Any
      }
        return [rcProvider, rcInterop]
    }

    // MARK: - Remote Config Interop Protocol
    func registerRolloutsStateSubscriber(subscriber: FIRRolloutsStateSubscriber, for namespace: String) {
        let instance : FIRRemoteConfig? = self.remoteConfigForNamespace(firebaseNamespace: namespace)
      guard let instance = instance else {
        return
      }
        [instance addRemoteConfigInteropSubscriber:subscriber];
    }
    
}
